src/main/java/org/mozilla/jss/pkcs11/PK11Signature.java

```diff
@@ -200,29 +209,43 @@ public void engineUpdate(byte b)
     public void engineUpdate(byte[] b, int off, int len)
         throws SignatureException, TokenException
     {
-        assert(b != null);
-        if( (state==SIGN || state==VERIFY) ) {
-            if(!raw && sigContext==null) {
-                throw new SignatureException("Signature has no context");
-            } else if( raw && rawInput==null) {
-                throw new SignatureException("Raw signature has no input stream");
-            }
-        } else {
-            assert(state == UNINITIALIZED);
-            throw new SignatureException("Signature is not initialized");
+        if (b == null) {
+            throw new SignatureException("No byte provided");
         }
-        assert(token!=null);
-        assert(tokenProxy!=null);
-        assert(algorithm!=null);
-        assert(key!=null);
-
-        if( raw ) {
+        validateUpdate();
+        if (raw) {
             rawInput.write(b, off, len);
         } else {
             engineUpdateNative( b, off, len);
         }
     }
 
+    private void validateUpdate() throws SignatureException, TokenException {
+        if (state == SIGN || state == VERIFY) {
+            if (!raw && sigContext == null) {
+                throw new SignatureException(NO_CONTEXT);
+            } else if (raw && rawInput == null) {
+                throw new SignatureException(NO_INPUT_STREAM);
+            }
+        } else {
+            if (state == UNINITIALIZED) {
+                throw new SignatureException(SIG_NOT_INITIALIZED);
+            }
+        }
+        if (token == null) {
+            throw new TokenException(NO_TOKEN);
+        }
+        if (tokenProxy == null) {
+            throw new TokenException(NO_TOKEN_PROXY);
+        }
+        if (algorithm == null) {
+            throw new SignatureException(NO_ALGORTIHM);
+        }
```

Comment:
I think it's not necessary to check `token` and `algorithm` here since they are set in the constructor and never change. You might want to move the `tokenProxy` check into the constructor as well.

Reply:
Ack, `tokenProxy` check moved into the constructor.